File: 프로그래머스/lv3/12927. 야근 지수/야근 지수.js
function solution(n, works) {
    var answer = 0;
    
    const l = works.length
    
    works.sort((a,b)=>b-a)
    while (n > 0) {

        let maxV = works[0]
        if (maxV === 0) return 0
        for(let i=0; i<l; i++) {
            if (works[i] === maxV) {
                works[i]--
                n--
                if (!n) break
            }
        }
        
        maxV --
        if (!maxV) break

    }
    
    for (const work of works) {
        answer += work**2
    }
    
    return answer;
}

File: 프로그래머스/lv3/12927. 야근 지수/야근 지수.py
import heapq

def solution(n, works):
    answer = 0
    max_heap = []

    for work in works:
        heapq.heappush(max_heap, -work)

    while n > 0:
        max_work = -heapq.heappop(max_heap)
        if max_work == 0:
            return 0

        max_work -= 1
        n -= 1

        heapq.heappush(max_heap, -max_work)

    while max_heap:
        work = -heapq.heappop(max_heap)
        answer += work ** 2

    return answer
